pkg/controller/useraccount/useraccount_controller.go

```diff
@@ -98,75 +99,74 @@ func (r *ReconcileUserAccount) Reconcile(request reconcile.Request) (reconcile.R
+func (r *ReconcileUserAccount) ensureUser(logger logr.Logger, userAcc *toolchainv1alpha1.UserAccount) (*userv1.User, bool, error) {
 	user := &userv1.User{}
 	if err := r.client.Get(context.TODO(), types.NamespacedName{Name: userAcc.Name}, user); err != nil {
 		if errors.IsNotFound(err) {
-			log.Info("creating a new user", "name", userAcc.Name)
-			user = newUser(userAcc)
-			if err := controllerutil.SetControllerReference(userAcc, user, r.scheme); err != nil {
+			logger.Info("creating a new user", "name", userAcc.Name)
+			if err := r.updateStatus(userAcc, toolchainv1alpha1.StatusProvisioning, err.Error()); err != nil {
 				return nil, false, err
 			}
-			if err := r.client.Create(context.TODO(), user); err != nil {
-				return nil, false, errs.Wrapf(err, "failed to create user '%s'", userAcc.Name)
+			user = newUser(userAcc)
+			err = controllerutil.SetControllerReference(userAcc, user, r.scheme)
+			if err == nil {
+				err = r.client.Create(context.TODO(), user)
+			}
+			if err != nil {
+				return nil, false, r.wrapErrorWithStatusUpdate(logger, userAcc, err, "failed to create user '%s'", userAcc.Name)
 			}
-			log.Info("user created successfully", "name", userAcc.Name)
+			logger.Info("user created successfully", "name", userAcc.Name)
 			return user, true, nil
 		}
-		return nil, false, errs.Wrapf(err, "failed to get user '%s'", userAcc.Name)
+		return nil, false, r.wrapErrorWithStatusUpdate(logger, userAcc, err, "failed to get user '%s'", userAcc.Name)
 	}
-	log.Info("user already exists", "name", userAcc.Name)
+	logger.Info("user already exists", "name", userAcc.Name)
 	return user, false, nil
 }
 
-func (r *ReconcileUserAccount) ensureIdentity(userAcc *toolchainv1alpha1.UserAccount) (*userv1.Identity, bool, error) {
+func (r *ReconcileUserAccount) ensureIdentity(logger logr.Logger, userAcc *toolchainv1alpha1.UserAccount) (*userv1.Identity, bool, error) {
 	name := getIdentityName(userAcc)
 	identity := &userv1.Identity{}
 	if err := r.client.Get(context.TODO(), types.NamespacedName{Name: name}, identity); err != nil {
 		if errors.IsNotFound(err) {
-			log.Info("creating a new identity", "name", name)
-			identity = newIdentity(userAcc)
-			if err := controllerutil.SetControllerReference(userAcc, identity, r.scheme); err != nil {
+			logger.Info("creating a new identity", "name", name)
+			if err := r.updateStatus(userAcc, toolchainv1alpha1.StatusProvisioning, err.Error()); err != nil {
 				return nil, false, err
 			}
-			if err := r.client.Create(context.TODO(), identity); err != nil {
-				return nil, false, errs.Wrapf(err, "failed to create identity '%s'", name)
+			identity = newIdentity(userAcc)
+			err = controllerutil.SetControllerReference(userAcc, identity, r.scheme)
+			if err == nil {
+				err = r.client.Create(context.TODO(), identity)
+			}
+			if err != nil {
+				return nil, false, r.wrapErrorWithStatusUpdate(logger, userAcc, err, "failed to create identity '%s'", name)
```

Comment:
This doesn't look obvious error_handling.  In case `SetControllerReference` returns error.  First we are testing `err == nil` and then we are testing `err != nil`.  It's testing err values two times instead of one.
So overall, code looks fine but execution suffers here.

Reply:
Yes, it’s because err can become not nil after the first check. The code is a bit ambiguous I agree. Alternatively we could check two errors independently which will make the code a bit more verbose but could make it easier to read.

Reply:
This is less go_ideomatic way. go says first check error and then proceed.  I really need to execute the code in my mind to understand it.  With that input, I will leave it up to you :+1:

Reply:
Addressed in https://github.com/codeready-toolchain/member-operator/pull/25/commits/81f9d57c0d73f11d0430e1ae57b28891038d1a96